Kz/gradient descent

KZhu-ME opened the pull request:
Utility that uses inverse exponential and linear regression to predict particle densities in a run to steady state simulation.

This is a plot of the test added.
![ar_ionxavg_start-50ns_interval-50ns_prediction-350ns](https://user-images.githubusercontent.com/86268612/166842313-35b56842-3f4e-4d50-b59b-e88e37ff076f.png)

KZhu-ME committed:
Added ArbitraryDistributionVolumeEmitter

KZhu-ME committed:
Updated particle weight calculation

KZhu-ME committed:
renamed thermionic_emission test directory to emission

KZhu-ME committed:
arbitrary seed from uniform distribution

KZhu-ME committed:
Bug fixes and changes to test

KZhu-ME committed:
add test for arbitrary seeding

KZhu-ME committed:
Changed position sampling, added test

KZhu-ME committed:
Removed unused imports and code

KZhu-ME committed:
revert traceback printout

KZhu-ME committed:
Added comments for binning particle positions

KZhu-ME committed:
increased rtol for test

KZhu-ME committed:
Allow input of any 2d array, add interpolation of 2d array onto current simulation grid

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

KZhu-ME committed:
Merge branch 'memaster' of github.com:ModernElectron/WarpX into kz/arbitrary_seed

KZhu-ME committed:
Updated changelog

KZhu-ME committed:
Merge branch 'kz/arbitrary_seed' of github.com:ModernElectron/WarpX into kz/arbitrary_seed

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

KZhu-ME committed:
Updated changelog

KZhu-ME committed:
Merge branch 'kz/arbitrary_seed' of github.com:ModernElectron/WarpX into kz/arbitrary_seed

KZhu-ME committed:
update version

KZhu-ME committed:
created steady_state_descent.py file

KZhu-ME committed:
Added coarsening algorithm

KZhu-ME committed:
Added ability to read numpy data and coarsen grid

KZhu-ME committed:
Added regression function calculation

KZhu-ME committed:
Solve system of equations rather than curve fit

KZhu-ME committed:
Use skimage block_reduce

KZhu-ME committed:
Hard coded inverse exponential solve

KZhu-ME committed:
take mean if exponential equation can't be solved

KZhu-ME committed:
Added special cases to use linear regression

KZhu-ME committed:
Merge branch 'memaster' of github.com:ModernElectron/WarpX into kz/gradient_descent

KZhu-ME committed:
Update docstring

KZhu-ME committed:
Added command line arguments

KZhu-ME committed:
Modified linear regression logic and added main function

KZhu-ME committed:
Added entry point

KZhu-ME committed:
Change output file name

KZhu-ME committed:
Change output file name

KZhu-ME committed:
Fixed interpolate grid

KZhu-ME committed:
updated arbitrary distribution emitter test

KZhu-ME committed:
Fix interpolation logic

KZhu-ME committed:
Changed to take in exact prediction timestep rather than a multiple

KZhu-ME committed:
Added ability to specify save directory

KZhu-ME committed:
Updated docstrings, removed unneeded files

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

KZhu-ME committed:
Added gradient descent test, refactored code

KZhu-ME committed:
Merge branch 'kz/gradient_descent' of github.com:ModernElectron/WarpX into kz/gradient_descent

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

KZhu-ME committed:
Changed test parameters

KZhu-ME committed:
Merge branch 'kz/gradient_descent' of github.com:ModernElectron/WarpX into kz/gradient_descent

KZhu-ME committed:
Merge branch 'memaster' of github.com:ModernElectron/WarpX into kz/gradient_descent

KZhu-ME committed:
Increase rtol of arbitrary seed test by 0.01

peterscherpelz commented on mewarpx/setup.py:
```diff
@@ -6,12 +6,12 @@
 
 requires = [
-    "numpy==1.20", # numba v0.54 requires numpy < 1.21.0
+    "numpy<1.22",  # numba v0.54 requires numpy < 1.21.0
```
Comment is out-of-date now

peterscherpelz commented on mewarpx/mewarpx/utils_store/plasma_density_oracle.py:
```diff
@@ -0,0 +1,221 @@
+"""
+This script reads in plasma densities at 3 time points and fits an exponential
+regression to predict the plasma density at a future time point. This script
```
Rephrase "exponential regression" to be more accurate

peterscherpelz replied:
In that it can be linear, exponential or mean

RemiLehe committed:
Apply shift in BTD for zmax different than 0 (#3091)

* Include shift in t_lab when `zmax` is different than 0

* Revert "Include shift in t_lab when `zmax` is different than 0"

This reverts commit 5f786f26ada3d9713221df653e28923557b5b614.

* Apply shift in BTD for zmax > 0

roelof-groenewald commented on mewarpx/mewarpx/utils_store/plasma_density_oracle.py:
```diff
@@ -0,0 +1,221 @@
+                            m, b = self._solve_linear(x_data, y_data)
+                            # time_skip is at x = 1, due to scaling down
+                            prediction = m + b
+                        # if data is linear or density increase is accelerating
```
This is a typo, right?
```suggestion
                        # if density is decelerating
```

KZhu-ME replied:
Yep

roelof-groenewald reviewed (commented):
Overall it looks good to me. I wonder though if we should have the input to the oracle be relative step numbers rather than absolute ones. The reason is that if we run a continuous cycle of sim - predict - seed - sim - predict - etc, the input script would need to know at what step the cycle currently is. Instead the oracle could look for density files and use the last 3 files (or last three with a given step spacing). Alternatively we just need to make the driving script smart enough to pick out the appropriate steps. I think it would be good to have an example driving script included in this PR.

KZhu-ME committed:
Update mewarpx/mewarpx/utils_store/plasma_density_oracle.py

Co-authored-by: Roelof Groenewald <40245517+roelof-groenewald@users.noreply.github.com>

KZhu-ME commented:
The reason for setting the prediction step as an absolute number is to allow more flexibility in deciding what the prediction step should be, and to make the calculation easier, (ie. it's easier to calculate what step is 5us rather than what multiple of the `interval` is at 5us). Additionally, since each new cycle of seed sim and run is a new simulation, the step count is reset to 0 at the start of each cycle, so the `predict_step` can remain the same for each cycle if you want to predict the same amount of time in the future.

I'll include the batch script I used in the corresponding mwxp PR. [34838700](https://github.com/ModernElectron/mwxp/pull/47)

KZhu-ME committed:
Update changelog

KZhu-ME committed:
Merge branch 'kz/gradient_descent' of github.com:ModernElectron/WarpX into kz/gradient_descent

ax3l committed:
AMReX: Weekly Update (#3095)

KZhu-ME commented:
Holding off from merging now, to add control script that will automatically choose prediction parameters.

KZhu-ME committed:
Update version

KZhu-ME committed:
Add controller for plasma density oracle

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lucafedeli88 committed:
Fix miscellaneous issues found with clang tidy (#3083)

* fix miscellaneous issues found with clang tidy
* fix miscellaneous issues found with clang tidy
* moved default constructors and destructors to header
* fix issue
* Use default again

lucafedeli88 committed:
Add a python script to generate timestep duration plots (#3036)

* add script to generate timestep duration plots
* disable transparent background
* fix bugs
* fix bug in doc
* Update plot_timestep_duration.rst
* add permissions
* using argparse
* using if __name__ == __main__:
* add page in workflows
* Update Docs/source/usage/workflows.rst
* added missing file
* Formatting

RemiLehe committed:
Apply shift for zmax>0 in old BTD (#3094)

* Apply shift for zmax>0 in old BTD
* BTD: also shift t_slice_lab

lucafedeli88 committed:
Use uniform error message formatting for all the amrex::Abort() calls in WarpX.cpp (#2952)

* use WARPX_ALWAYS_ASSERT_WITH_MESSAGE instead of Abort()
* fix bug
* Update Source/WarpX.cpp
* fixed bug
* fixed bug
* Update Source/WarpX.cpp
* Update Source/WarpX.cpp
Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

KZhu-ME committed:
WIP oracle control

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

amadou38 committed:
Use queryWithParser to Parse v_particle_pml (#3096)

Use queryWithParser, instead of query, to parse v_particle_pml, in order to support functions like sin, cos, sqrt, etc.

KZhu-ME committed:
Allow oracle control bins to be overridden

KZhu-ME committed:
Merge branch 'kz/gradient_descent' of github.com:ModernElectron/WarpX into kz/gradient_descent

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

peterscherpelz commented on mewarpx/mewarpx/utils_store/util.py:
```diff
@@ -402,3 +402,17 @@ def interpolate_from_grid(coords, grid):
     return fpos
+
+
+def mwx_round(x, decimals=0, base=1):
+    """Rounding function that allows rounding around a base value.
+    Arguments:
+        x (float): value to round
+        decimals (int): number of decimal places to round to, defaults to 0.
+            If negative specifies positions left of the decimal point.
+        base (float): base number to round around
+
+    Returns:
+        val (Union[int, float]): rounded value
+    """
+    return np.round(base * np.round(x / base), decimals=decimals)
```
If I'm reading this right you're trying to get a multiple of `base`. I don't really understand why `return base * np.round(x / base)` is insufficient in that case. As it is, if `base << 1`, as it is when used in this PR, I find it non-intuitive that this will frequently round to 0 if `decimals` is incorrect.

peterscherpelz commented on mewarpx/mewarpx/utils_store/oracle_control.py:
```diff
@@ -0,0 +1,144 @@
+        self.output_file = os.path.join(self.work_dir, "oracle_control.txt")
+
+        # remove output file if it exists so that the autorun script will exit
+        # if any unexpected erros are thrown in this script
```
```suggestion
        # if any unexpected errors are thrown in this script
```

peterscherpelz commented on mewarpx/mewarpx/utils_store/oracle_control.py:
```diff
@@ -0,0 +1,144 @@
+            self.latest_run = os.path.join(self.work_dir, f"{self.prefix}_0")
+            mwxutil.mkdir_p(self.latest_run)
+
+        # check if latest run is completed successfully
```
```suggestion
        # check if latest run completed successfully
```

peterscherpelz commented on mewarpx/mewarpx/utils_store/plasma_density_oracle.py:
```diff
@@ -0,0 +1,221 @@
+                regression
+            interval (int): number of time steps between data points
+            predict_step (int): time step of prediction
+            species_names (list): list of species names or identifying shared
```
Both here and in the OracleControl, either this docstring or the main docstring should clearly describe how the shared substring works. It's otherwise not clear that 'Xe_ion' will always grab 'Xe_ion' and 'Xe_wire' IMO.
A longer-term-solution to consider might actually be to allow the full glob syntax in these strings, or at least the `*` which you could strip out for the purpose of saving files.

peterscherpelz reviewed (commented):
I hope these comments help. I didn't thoroughly work to understand the workflow, so I'll probably have adjustments to ask for in the future... but no reason to hold up the PR at this point.

peterscherpelz commented:
(The above said, I'm really excited about this! Mostly just feeling guilty I don't have more time to review right now, not worried about what's here at all 👍 )

EZoni committed:
Fix Wrong Abort Message for PSATD w/ Time Averaging (#3098)

ax3l committed:
Doc: Improve Doxygen Links (#3099)

* Doc: Improve Doxygen Links

- download also openPMD-api tag files & link them
- improve developer page: clearer links

* Update: .gitignore

KZhu-ME committed:
Added command line arguments

dpgrote committed:
Fix Mirrors for RZ (NullifyMF) (#3101)

KZhu-ME committed:
Fixed bugs, added info logging

KZhu-ME committed:
added ability to modify run_script

KZhu-ME committed:
removed unnecessary code

KZhu-ME committed:
Cleaned up code

KZhu-ME committed:
Merge branch 'memaster' of github.com:ModernElectron/WarpX into kz/gradient_descent

ax3l committed:
Fix Breaking Pip 22.1 Logic (#3106)

* CI: Fix Windows

Pip's `--no-build-isolation` says we have to have build-deps pre-installed.
But Pip does offer no option to install those from `pyproject.toml`.

* Pip 22.1: in-tree-build default

And dropped the old option with a fatal error... classic.

* Pip no-build isolation: pre-install build tools

not installed in non-isolated builds
- we want isolated builds, because they have a derministic
  path that we can cache with ccache

lucafedeli88 committed:
fix bug (#3110)

EZoni committed:
Fix Laser Injection in RZ (#3104)

lgiacome committed:
Add STL files support in pywarpx (#3089)

* Added STL files support in pywarpx

* Add new EB bucket

* Added name to file headers

* Fixing year in  file headers

* Checking that stl file and imp function are not both specified

* Renamed EB bucket to EB2

* Adding STL documentation

* Implement suggestions from code review

Co-authored-by: lgiacome <lorenzo.giacome@cern.ch>

EZoni committed:
Stricter Tolerance for 3D Langmuir Tests (#3097)

ax3l committed:
CI: HIP with https (#3111)

Small update in HIP docs: now supports HTTPS:
https://rocmdocs.amd.com/en/latest/Installation_Guide/Installation-Guide.html#ubuntu

EZoni committed:
Add Headers to Input File Written by PICMI (#3093)

* Add Headers to Input File Written by PICMI
* Commit Suggestion by @dpgrote
* Add Comment
* Add Empty Line
* Slightly Better Readable with a newline

KZhu-ME committed:
Refactored code to be clearer

ax3l committed:
AMReX/PICSAR: Weekly Update (#3112)

* AMReX: Weekly Update

* PICSAR: Weekly Update

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#3113)

updates:
- [github.com/Lucas-C/pre-commit-hooks: v1.1.13 → v1.1.14](https://github.com/Lucas-C/pre-commit-hooks/compare/v1.1.13...v1.1.14)

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

lucafedeli88 committed:
Use uniform error message formatting in more source files (#2964)

* use WARPX_ALWAYS_ASSERT_WITH_MESSAGE instead of Abort()

* fix bug

* Update Source/WarpX.cpp

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

* use more uniform text formatting

* fixed bug

* fixed bug

* using Utils::TextMsg in more files

* fixed bug

* use Utils::TextMsg in more files

* fixed bug

* use Utils::TextMsg in more files

* fixed last merge conflict

* fixed bug

* fix indentation

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

PhilMiller committed:
Avoid resizing ParticleBoundaryBuffer allocation to total particle count (#3068)

* Avoid resizing ParticleBoundaryBuffer allocation to total particle count

* Switch atomic increments over to reductions

ax3l committed:
Lassen (LLNL): ADIOS2, BLAS++ & LAPACK++ (#3071)

* Lassen (LLNL): ADIOS2, BLAS++ & LAPACK++

Document how to install ADIOS2 (for openPMD `.bp`) and
BLAS++/LAPACK++ (for PSATD in RZ) on Lassen at LLNL.

* Update: ADIOS2~SST, BLAS++/LAPACK++ w/ ESSL

atmyers committed:
Add workaround for Libfabric bug on Crusher to job script. (#3114)

* Add workaround for Libfabric bug on Crusher to job script.

* add known issues section for Crusher

KZhu-ME committed:
Update docstrings

KZhu-ME committed:
Added test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

KZhu-ME committed:
Added documentation about oracle_control

KZhu-ME committed:
Merge branch 'kz/gradient_descent' of github.com:ModernElectron/WarpX into kz/gradient_descent

KZhu-ME committed:
Updated changelog

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

KZhu-ME committed:
Fixed TOTAL_TIME calculation

KZhu-ME committed:
Merge branch 'kz/gradient_descent' of github.com:ModernElectron/WarpX into kz/gradient_descent

ax3l committed:
ABLASTR: localIDtoGlobal (#3115)

* ABLASTR: localIDtoGlobal

Move the control of `localIDtoGlobal` to ABLASTR for particle IDs.

* WarpX: Use ABLASTR now

roelof-groenewald committed:
Update ME fork with latest upstream/developement and test build image

KZhu-ME committed:
update wheel in docker build

KZhu-ME committed:
build image

KZhu-ME committed:
Merge branch 'upstream_update' of github.com:ModernElectron/WarpX into kz/gradient_descent

KZhu-ME committed:
change numpy to oldest-supported-numpy for docker

KZhu-ME committed:
build image

peterscherpelz commented on mewarpx/changelog.csv:
```diff
@@ -1,12 +1,30 @@
 - Added
   :class:`mewarpx.utils_store.plasma_density_oracle.PlasmaDensityOracle` to
   predict particle densities at a given time point based on 3 input data points.
+- Added :class:`mewarpx.utils_store.plasma_density_oracle.PlasmaDensityOracle`
```
I think you're referencing the wrong class here?

peterscherpelz commented on mewarpx/mewarpx/utils_store/oracle_control.py:
```diff
@@ -1,85 +1,112 @@
+"""
```
This file has lots of lines that overflow 80 characters some. Probably you just had your text environment set a bit off?

peterscherpelz commented on mewarpx/mewarpx/utils_store/oracle_control.py:
```diff
@@ -1,85 +1,112 @@
+        self.start_step = self._time_to_step(first_data)
 
         # set class variables
+        if run_script is None:
+            run_script = "run_simulation.py"
+        if prefix is None:
+            prefix = "start"
         if species_names is None:
             species_names = ["electron", "ion"]
+        if coarsening_factor is None:
+            coarsening_factor = 8
```
I assume defaults are in here rather than function definition because the command line calls explicitly include None in some cases. That's fine, but a comment here would be helpful.
Also, add the defaults to the docstring for each argument; otherwise documentation won't show defaults at all.

peterscherpelz commented on mewarpx/mewarpx/utils_store/oracle_control.py:
```diff
@@ -132,13 +195,86 @@ def _time_to_step(self, time):
+         'save_dir': None,
+         'species_names': None,
+         'work_dir': None}
+    controller = OracleControl(**a)
```
Should `controller.main()` be run here too?

peterscherpelz commented on mewarpx/mewarpx/utils_store/plasma_density_oracle.py:
```diff
@@ -1,10 +1,11 @@
 """
```
Same 80-char line limit concern here

peterscherpelz reviewed (changes requested):
Thanks! A number of small questions but overall looking great.

KZhu-ME committed:
Updated formatting and documentation

peterscherpelz reviewed (approved):
lgtm!
